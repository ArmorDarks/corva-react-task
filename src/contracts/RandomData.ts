// @todo That file isn't used by server, be careful

export const RANDOM_DATA_CHANNEL = 'data'

export type Timestamp = number
export type RandomDataValue = number

/**
 * Response from the server
 */
export interface RandomData {
  value: RandomDataValue
  timestamp: Timestamp
}
