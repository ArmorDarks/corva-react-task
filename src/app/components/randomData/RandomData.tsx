/* eslint no-unused-vars: "off" */
/* eslint @typescript-eslint/no-unused-vars: "error" */

import React from 'react'

import { BarChart, SplineChart } from '..'
import { RandomData as RandomDataContract } from '../../../contracts/RandomData'
import { Threshold, updateThreshold } from '../../store/randomData'

const CHART_MAX_SERIES = 10

export interface Props {
  isConnected?: boolean,
  randomData: readonly RandomDataContract[],
  threshold?: Threshold,
  updateThreshold: typeof updateThreshold,
}

/**
 * Display Random Data charts and some additional controls
 */
const RandomData: React.FC<Props> = ({ isConnected, randomData, threshold, updateThreshold }) => {
  return (
    <div className='h-margin-top+'>

      <div className='Wrapper h-flex h-flex--wrap h-flex-x--between h-flex-y--center h-margin-ends+'>

        <div className='h-flex h-flex--wrap h-flex-y--center'>
          <h1 className='h-inline h-margin-ends'>Random Data Stats</h1>

          <div className='h-margin-left h-margin-ends'>
            {isConnected ? 'online' : 'offline'}
          </div>

        </div>

        <div className='h-margin-ends'>
          <label className='h-text- h-margin-right-' htmlFor='threshold'>Threshold</label>
          <input
            type='number'
            id='threshold'
            name='threshold'
            onChange={(e) => updateThreshold(e.target.value === '' ? null : +e.target.value)}
            value={threshold === null ? '' : threshold}
          />
        </div>

      </div>

      <div className='o-grid'>

        <div className='o-grid__iterm h-1/2'>
          <BarChart
            seriesName='Random data occurrences'
            seriesData={randomData.map(({ value }) => value)}
          />
        </div>

        <div className='o-grid__iterm h-1/2'>
          <SplineChart
            seriesName='Random data'
            seriesData={randomData.map((data) => [data.timestamp, data.value] as const)}
            threshold={threshold === null ? undefined : threshold}
            limit={CHART_MAX_SERIES}
          />
        </div>

      </div>

    </div>
  )
}

export default RandomData
