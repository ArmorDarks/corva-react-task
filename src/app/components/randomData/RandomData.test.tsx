/* eslint-env jest */
/* eslint no-unused-vars: "off" */
/* eslint @typescript-eslint/no-unused-vars: "error" */

import React from 'react'
import { shallow } from 'enzyme'

import RandomData from './RandomData'

describe('RandomData', () => {
  it('should render', () => {
    const Component = () =>
      <RandomData
        isConnected
        randomData={[]}
        updateThreshold={(() => {}) as any}
      />

    expect(shallow(Component())).toMatchSnapshot()
  })

  it('should render with data', () => {
    const Component = () =>
      <RandomData
        isConnected
        randomData={[{
          value: -59.168341304150964,
          timestamp: 1570474691564
        }, {
          value: -81.39908194859946,
          timestamp: 1570474694403
        }]}
        updateThreshold={(() => {}) as any}
      />

    expect(shallow(Component())).toMatchSnapshot()
  })

  it('should update threshold', () => {
    const mockUpdateThreshold = jest.fn()

    const component = shallow(
      <RandomData
        isConnected
        randomData={[{
          value: -59.168341304150964,
          timestamp: 1570474691564
        }]}
        threshold={15}
        updateThreshold={mockUpdateThreshold}
      />
    )

    const $input = component.find('input')

    const value = 23
    $input.simulate('change', { target: { value } })

    expect(mockUpdateThreshold.mock.calls).toEqual([[value]])
  })
})
