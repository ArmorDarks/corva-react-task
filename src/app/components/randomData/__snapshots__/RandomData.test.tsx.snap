// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RandomData should render 1`] = `
<div
  className="h-margin-top+"
>
  <div
    className="Wrapper h-flex h-flex--wrap h-flex-x--between h-flex-y--center h-margin-ends+"
  >
    <div
      className="h-flex h-flex--wrap h-flex-y--center"
    >
      <h1
        className="h-inline h-margin-ends"
      >
        Random Data Stats
      </h1>
      <div
        className="h-margin-left h-margin-ends"
      >
        online
      </div>
    </div>
    <div
      className="h-margin-ends"
    >
      <label
        className="h-text- h-margin-right-"
        htmlFor="threshold"
      >
        Threshold
      </label>
      <input
        id="threshold"
        name="threshold"
        onChange={[Function]}
        type="number"
      />
    </div>
  </div>
  <div
    className="o-grid"
  >
    <div
      className="o-grid__iterm h-1/2"
    >
      <BarChart
        seriesData={Array []}
        seriesName="Random data occurrences"
        step={10}
      />
    </div>
    <div
      className="o-grid__iterm h-1/2"
    >
      <SplineChart
        limit={10}
        seriesData={Array []}
        seriesName="Random data"
      />
    </div>
  </div>
</div>
`;

exports[`RandomData should render with data 1`] = `
<div
  className="h-margin-top+"
>
  <div
    className="Wrapper h-flex h-flex--wrap h-flex-x--between h-flex-y--center h-margin-ends+"
  >
    <div
      className="h-flex h-flex--wrap h-flex-y--center"
    >
      <h1
        className="h-inline h-margin-ends"
      >
        Random Data Stats
      </h1>
      <div
        className="h-margin-left h-margin-ends"
      >
        online
      </div>
    </div>
    <div
      className="h-margin-ends"
    >
      <label
        className="h-text- h-margin-right-"
        htmlFor="threshold"
      >
        Threshold
      </label>
      <input
        id="threshold"
        name="threshold"
        onChange={[Function]}
        type="number"
      />
    </div>
  </div>
  <div
    className="o-grid"
  >
    <div
      className="o-grid__iterm h-1/2"
    >
      <BarChart
        seriesData={
          Array [
            -59.168341304150964,
            -81.39908194859946,
          ]
        }
        seriesName="Random data occurrences"
        step={10}
      />
    </div>
    <div
      className="o-grid__iterm h-1/2"
    >
      <SplineChart
        limit={10}
        seriesData={
          Array [
            Array [
              1570474691564,
              -59.168341304150964,
            ],
            Array [
              1570474694403,
              -81.39908194859946,
            ],
          ]
        }
        seriesName="Random data"
      />
    </div>
  </div>
</div>
`;
